### PR TITLE
Added Azure.MySQL.UseFlexible

### LINKED ISSUE #1841: Azure Database for MySQL Single Servers should consider to upgrade to Azure Database for MySQL Flexible Servers
# Rule request

## Suggested rule change

Azure Database for MySQL Single Servers are on the retirement path.

We should recommend customers to upgrade their single servers to the flexible servers variant which also will offers new features, functions and security.

image

Operational excellence pillar suits this rule.

## Applies to the following

The rule applies to the following:

- Resource type: **[Microsoft.DBforMySQL/servers]**

## Sample data
[What's happening to Azure Database for MySQL - Single Server?](https://learn.microsoft.com/azure/mysql/single-server/whats-happening-to-mysql-single-server)
[Template reference](https://learn.microsoft.com/en-us/azure/templates/microsoft.dbformysql/flexibleservers)

### PR DESCRIPTION
**Pull request by bengeset96**
## PR Summary

Fixes #1841 

Added Azure.MySQL.UseFlexible rule.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by bengeset96** [comment]
@BernieWhite 

I want to merge main into this branch after we have the new updates to the `GeoRedundantBackup` rule as it has some dependencies.

The idea here is:

- We include flexible deployment model so the customer see that he passed the rule.
- Run the rule with `-Level Warning` so it does not produce terminating error because I can't really see that the single server deployment model has been given any last date. Maybe change it to default behaviour when we see that there is an date settled for the fully retirement.

File: docs/en/rules/Azure.MySQL.UseFlexible.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+## SYNOPSIS
+
+Use Azure Database for MySQL Flexible Server deployment model.
+
+## DESCRIPTION
+
+Azure Database for MySQL Single Server is on the retirement path. Upgrade to Azure Database for MySQL Flexible Server.
+
+Azure Database for MySQL Flexible Server to bring you the best of Azure’s open-source database platform.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.MySQL.UseFlexible.md:19]:
```suggestion
Azure Database for MySQL Flexible Server provides additional options for resilience and scalability above the Single Server deployment model.
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Thanks @bengeset96. There are a few differences so I agree we shouldn't make that break pipelines yet and a warning is the best approach.

I've added a few comments to suggest alternatives for wording but let me know if you have an questions.

File: docs/en/rules/Azure.MySQL.UseFlexible.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+Check out the `NOTES` section for more details about the retirement path and migration steps.
```
[the remark this answers is not in the document]
Comment (bengeset96) [reply docs/en/rules/Azure.MySQL.UseFlexible.md]: Yes, this is just my mistake.

File: src/PSRule.Rules.Azure/rules/Azure.MySQL.Rule.ps1 [hunk 3]
```diff
@@ -55,6 +55,16 @@ Rule 'Azure.MySQL.GeoRedundantBackup' -Ref 'AZR-000323' -Type 'Microsoft.DBforMy
...
 }
 
+# Synopsis: Use Azure Database for MySQL Flexible Server deployment model.
+Rule 'Azure.MySQL.UseFlexible' -Ref 'AZR-000325' -Type 'Microsoft.DBforMySQL/flexibleServers', 'Microsoft.DBforMySQL/servers' -Level Warning -Tag @{ release = 'GA'; ruleSet = '2022_12'; } {
+    if ($PSRule.TargetType -eq 'Microsoft.DBforMySQL/flexibleServers') {
+        $Assert.Pass()
+    }
+    elseif ($PSRule.TargetType -eq 'Microsoft.DBforMySQL/servers') {
+        $Assert.Fail($LocalizedData.SingleDeploymentModelRetirement)
+    }
```
[the remark this answers is not in the document]
Comment (bengeset96) [reply src/PSRule.Rules.Azure/rules/Azure.MySQL.Rule.ps1:65]: @BernieWhite 

Great, I'll start to to use this approach more!

**Review by BernieWhite — APPROVED** [review]
Great work @bengeset96 ! All good to merge.

**Pull request merged**